src/current/v24.3/ldap.md

```diff
@@ -0,0 +1,206 @@
+host    all    root    0.0.0.0/0    password';
+~~~
+
+For more information on the HBA fields and HBA in CockroachDB, generally, refer to [SQL Authentication configuration](https://www.cockroachlabs.com/docs/stable/security-reference/authentication#authentication-configuration).
+
+Q: Should we specify any LDAP-specific recommendations/considerations for the first four fields (host, all, all, all) or enhance the authentication page to better cover them?
+
+#### Step 3: Configure TLS (Optional)
+
+If using LDAPS, configure the necessary certificates.
+
+Set a custom CA certificate if needed:
+
+~~~ sql
+    SET CLUSTER SETTING server.ldap_authentication.domain_ca = '<PEM_ENCODED_CA_CERT>';
+~~~
+
+Configure a client certificate for mTLS if required:
+
+~~~ sql
+    SET CLUSTER SETTING server.ldap_authentication.client.tls_certificate = '<PEM_ENCODED_CERT>';
+    SET CLUSTER SETTING server.ldap_authentication.client.tls_key = '<PEM_ENCODED_KEY>';
+~~~
+
+#### Step 4: Create Database Users
+
+Q: Confirm: Does a new customer need to create one user here or all users in bulk? 
+
+Create a CockroachDB user that matches the LDAP username:
+
+~~~ sql
+CREATE ROLE username LOGIN;
+~~~
+
+### Connecting with LDAP Authentication
+
+Users can connect using their LDAP credentials:
+
+~~~ bash
+# Method 1: Password in connection string
+cockroach sql --url "postgresql://username:ldap_password@host:26257" --certs-dir=certs
+
+# Method 2: Password in environment variable
+export PGPASSWORD='ldap_password'
+cockroach sql --url "postgresql://username@host:26257" --certs-dir=certs
+~~~
+
+## LDAP Authorization
+
+### Overview
+
+LDAP authorization allows CockroachDB to use LDAP/AD groups as the source of truth for [role](https://www.cockroachlabs.com/docs/stable/security-reference/authorization#roles) assignments. When enabled:
+
+1. During login, CockroachDB fetches the user's group memberships from LDAP.
+2. Groups are mapped to CockroachDB roles using the group's Common Name (CN).
+3. The user is automatically granted membership to corresponding roles.
+4. Any existing role memberships not matched to LDAP groups are revoked.
+
+### Configuration
+
+#### Step 1: Enable LDAP Authorization
+
+Add the `ldapgrouplistfilter` parameter to your HBA configuration:
+
+~~~ sql
+SET CLUSTER SETTING server.host_based_authentication.configuration = '
+host    all    all    all    ldap    ldapserver=ldap.example.com 
+    ldapport=636 
+    "ldapbasedn=ou=users,dc=example,dc=com" 
+    "ldapbinddn=cn=readonly,dc=example,dc=com" 
+    ldapbindpasswd=readonly_password 
+    ldapsearchattribute=uid 
+    "ldapsearchfilter=(memberof=cn=cockroachdb_users,ou=groups,dc=example,dc=com)"
+    "ldapgrouplistfilter=(objectClass=groupOfNames)"
+host    all    root    0.0.0.0/0    password';
+~~~
+
+The `ldapgrouplistfilter` parameter defines which LDAP objects should be considered as groups.
+
+Q: Get more specific/realistic example from Eng?
```

Comment:
The operator should restrict the scope of groups being fetched as unintentionally a lot of groups for which sql user is a member of will be fetched and assigned on CRDB. to restrict groups being fetched the grouplistfilter option will be as follows:
```
ldapgrouplistfilter="(|(cn=ldap_user_parent_1)(cn=ldap_user_parent_2))"
```
where only groups having common names ldap_user_parent_1 and ldap_user_parent_2 are fetched. We should also be creating the corresponding roles for them on CRDB.

Reply:
> you're saying that we're specifying specific groups to fetch here

Yes these are the roles that can get potentially fetched if they are assigned to the ldap user attempting login

> we will assign a matching role to cockroachdb users who are part of that group

We will assign a matching role for each group that got fetched from AD for the user

> but first, the customer needs to manually create all the matching role names in CockroachDB

Yes customer need to create the group roles first which they intend to sync. These group roles are actually crdb roles with rolename="group common name(cn)"

> Do we only assign the roles (and update them) to a user when a given user authenticates

The group roles and their privileges stay consistent across login attempt. Only the group role assignments for the user change depending on which groups were fetched from AD and their assignment on crdb.
The `SHOW roles` stmt shows the users and the group role assignments
```
root@localhost:26257/defaultdb> SHOW roles;
               username              | options |              member_of
-------------------------------------+---------+---------------------------------------
  admin                              |         | {}
  crl-crlcloud-dev-domain-sync-users | NOLOGIN | {}
  crl-dev-test                       | NOLOGIN | {}
  root                               |         | {admin}
  security_admin                     | NOLOGIN | {}
  sourav.sarangi                     |         | {admin}
  souravcrl                          |         | {}
  souravpwd                          |         | {}
  testuser                           |         | {}
  testuserdemo                       |         | {crl-crlcloud-dev-domain-sync-users}
(10 rows)

Time: 18ms total (execution 17ms / network 1ms)
```

Reply:
The `show sessions` now  also shows the authenticating method:
```
root@localhost:26257/defaultdb> show sessions;
  node_id |            session_id            | status |  user_name   | authentication_method | client_address  | application_name |    active_queries     | last_active_query |         session_start         |      active_query_start       | num_txns_executed | trace_id | goroutine_id
----------+----------------------------------+--------+--------------+-----------------------+-----------------+------------------+-----------------------+-------------------+-------------------------------+-------------------------------+-------------------+----------+---------------
        1 | 1807ce82f20f10c80000000000000001 | ACTIVE | root         | cert-password         | 127.0.0.1:58106 | $ cockroach sql  | SHOW CLUSTER SESSIONS | SHOW database     | 2024-11-14 10:20:43.549616+00 | 2024-11-15 11:52:42.887729+00 |                42 |        0 |         1747
        1 | 1807cf5e1034b0680000000000000001 | IDLE   | testuserdemo | ldap                  | 127.0.0.1:58667 | $ cockroach sql  |                       | SHOW database     | 2024-11-14 10:36:24.726124+00 | NULL                          |                 4 |        0 |        17237
(2 rows)

Time: 11ms total (execution 10ms / network 1ms)
```